Wikipedia/Code/WMFAppViewController+Extensions.swift

```diff
@@ -97,13 +101,119 @@ extension WMFAppViewController: NotificationsCenterPresentationDelegate {
+
+        while let presentedViewController = currentController?.presentedViewController {
+            if let presentedNavigationController = (presentedViewController as? UINavigationController) {
+                for viewController in presentedNavigationController.viewControllers {
+                    if let editingFlowViewController = viewController as? EditingFlowViewController {
+                        return editingFlowViewController
+                    }
+                }
+            } else if let editingFlowViewController = presentedViewController as? EditingFlowViewController {
+                return editingFlowViewController
+            }
+            
+            currentController = presentedViewController
+        }
+
+        return nil
+    }
+    
+    private var topMostViewController: UIViewController? {
+            
+        var topViewController: UIViewController = navigationController ?? self
+
+        while let presentedViewController = topViewController.presentedViewController {
+            topViewController = presentedViewController
+        }
+
+        return topViewController
     }
+    
+    private func presentEditorAlert(on viewController: UIViewController, confirmationBlock: @escaping () -> Void) {
+        
+        let title = CommonStrings.editorExitConfirmationTitle
+        let message = CommonStrings.editorExitConfirmationBody
+        
+        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
+        let discardAction = UIAlertAction(title: CommonStrings.discardEditsActionTitle, style: .destructive) { _ in
+            confirmationBlock()
+        }
+        let cancelAction = UIAlertAction(title: CommonStrings.cancelActionTitle, style: .cancel)
+        
+        alertController.addAction(discardAction)
+        alertController.addAction(cancelAction)
+        
+        viewController.present(alertController, animated: true, completion: nil)
+        
+    }
+}
+
+fileprivate extension UIViewController {
+    
+    /// Returns self or embedded view controller (if self is a UINavigationController) if conforming to NotificationsCenterFlowViewController
+    /// Does not consider presenting view controllers
+    var notificationsCenterFlowViewController: NotificationsCenterFlowViewController? {
+        
+        if let viewController = self as? NotificationsCenterFlowViewController {
+            return viewController
+        }
+        
+        if let navigationController = self as? UINavigationController,
+           let viewController = navigationController.viewControllers.last as? NotificationsCenterFlowViewController {
+            return viewController
+        }
+
+        return nil
+    }
+}
+
+
+/// View Controllers that have an editing element (Section editor flow, User talk pages, Article description editor)
+protocol EditingFlowViewController where Self: UIViewController {
+    var shouldDisplayAlert: Bool { get }
```

Comment:
Because all these interested objects are gonna inherit this property, I'd recommend we take a page out of Objective-C's book and be a bit more verbose about what we call this in an effort to 1. be more explicit about what it controls and 2. prevent unintended misuse of the property in future work in these controllers. 

Just a couple ideas: `shouldDisplayDiscardEditsAlert`, `shouldDisplayExitConfirmationAlert`, something like that?

Reply:
Good idea! I went with `shouldDisplayExitConfirmationAlert` for this